File: boj/python/dfs/boj1260.py
import sys
from collections import deque

dx = [0, 1, 0, -1]
dy = [1, 0, -1, 0]


def solution():
    visited = [False for _ in range(n + 1)]
    dfs_track = []
    bfs_track = []

    def dfs(now):
        visited[now] = True
        dfs_track.append(now)

        for w in graph[now]:
            if not visited[w]:
                # dfs_track.append(w)
                dfs(w)
        return

    def bfs(start):
        bfs_track.append(v)
        visited[v] = True

        q = deque()
        q.append(start)

        while q:
            now = q.popleft()
            for w in graph[now]:
                if not visited[w]:
                    q.append(w)
                    visited[w] = True
                    bfs_track.append(w)
        return

    dfs(v)

    visited = [False for _ in range(n + 1)]
    bfs(v)

    print(*dfs_track)
    print(*bfs_track)

    return


if __name__ == "__main__":
    input = sys.stdin.readline
    n, m, v = map(int, input().rstrip().split())
    graph = [[] for _ in range(n + 1)]

    for _ in range(m):
        v1, v2 = map(int, input().rstrip().split())
        graph[v1].append(v2)
        graph[v2].append(v1)

    for i in range(n + 1):
        graph[i].sort()

    solution()

File: src/brute_force/Boj1260.java
package brute_force;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Deque;
import java.util.List;
import java.util.StringTokenizer;

// (실버2) boj 1260 DFS와 BFS
public class Boj1260 {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		StringBuilder dfsResult = new StringBuilder();
		StringBuilder bfsResult = new StringBuilder();

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int v = Integer.parseInt(st.nextToken());

		List<List<Integer>> graph = new ArrayList<>();

		for (int i = 0; i < n + 1; i++) {
			graph.add(new ArrayList<>());
		}

		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine());

			int w = Integer.parseInt(st.nextToken());
			int u = Integer.parseInt(st.nextToken());

			graph.get(w).add(u);
			graph.get(u).add(w);
		}

		Deque<Integer> deque = new ArrayDeque<>();
		boolean[] visited = new boolean[n + 1];
		deque.offerLast(v);

		// dfs 탐색
		while (!deque.isEmpty()) {
			Integer now = deque.pollLast();
			if(visited[now]) continue;
			visited[now] = true;
			dfsResult.append(now).append(' ');

			graph.get(now).sort(Collections.reverseOrder());
			for (Integer next : graph.get(now)) {
				deque.offerLast(next);
			}
		}

		// 그래프 표현 방버 : 인저행렬, 인접리스트
		// 그ㅐ프 트리 정의
		// dfs, bfs 탐색

		visited = new boolean[n + 1];
		deque.offerLast(v);
		visited[v] = true;

		// bfs 탐색
		while (!deque.isEmpty()) {
			Integer now = deque.pollFirst();
			bfsResult.append(now).append(' ');

			graph.get(now).sort(Comparator.naturalOrder());
			for (Integer next : graph.get(now)) {
				if(visited[next]) continue;
				visited[next] = true;
				deque.offerLast(next);
			}
		}

		System.out.println(dfsResult);
		System.out.println(bfsResult);
	}
}
